### PR TITLE
Improve performance of block-based scheduling

### LINKED ISSUE #1457: Scheduling time grows exponentially with circuit depth
The execution time of a pass manager with scheduling and dynamical decoupling passes grows roughly exponentially with the circuit size. 

**Acceptance criteria**
Execution time should be approximately linear. 

**Details**

For example, schedule the following circuit with variable `N`
```python
from qiskit import transpile
from qiskit.transpiler.passmanager import PassManager
from qiskit.circuit.library import XGate, YGate
from qiskit import QuantumCircuit, QuantumRegister, ClassicalRegister
from qiskit_ibm_runtime.fake_provider.backends.sherbrooke import FakeSherbrooke
from qiskit_ibm_runtime.transpiler.passes.scheduling import (
    DynamicCircuitInstructionDurations,
    ALAPScheduleAnalysis,
    PadDynamicalDecoupling,
)

backend = FakeSherbrooke()
nqreg = 121
qreg = QuantumRegister(nqreg)
creg = ClassicalRegister(nqreg)
qc = QuantumCircuit(qreg, creg)
for qind in range(nqreg):
    qc.sx(qind)
qc.barrier()
for _ in range(N):
    qc.cx(qreg[:-1:2], qreg[1:-1:2])
    qc.barrier()
    qc.measure(qreg[::2], creg[::2])
    qc.barrier()
qc.measure(qreg, creg)

transpiled_circ = transpile(qc, backend)    
decoupling_seq = [XGate(), XGate()]
durations = DynamicCircuitInstructionDurations.from_backend(backend)
pm_dd = PassManager(
    [
        ALAPScheduleAnalysis(durations),
        PadDynamicalDecoupling(
            durations, decoupling_seq,
            skip_reset_qubits=False
        ),
    ]
)
```

e.g. with `N=6`:
```python
from pyinstrument import Profiler
with Profiler() as profiler:
     pm_dd.run(transpiled_circ)
```
```
64.999 <module>  ..\..\AppData\Local\Temp\ipykernel_23716\2696880545.py:1
└─ 64.998 wrapper  ..\..\GitHub\qiskit-terra\qiskit\transpiler\passmanager.py:417
   └─ 64.998 PassManager.run  ..\..\GitHub\qiskit-terra\qiskit\transpiler\passmanager.py:127
      └─ 64.998 PassManager.run  ..\..\GitHub\qiskit-terra\qiskit\passmanager\passmanager.py:170
         └─ 64.998 _run_workflow  ..\..\GitHub\qiskit-terra\qiskit\passmanager\passmanager.py:268
            └─ 64.664 FlowControllerLinear.execute  ..\..\GitHub\qiskit-terra\qiskit\passmanager\base_tasks.py:202
               └─ 64.664 ALAPScheduleAnalysis.execute  ..\..\GitHub\qiskit-terra\qiskit\transpiler\basepasses.py:189
                  └─ 64.664 ALAPScheduleAnalysis.execute  ..\..\GitHub\qiskit-terra\qiskit\passmanager\base_tasks.py:72
                     ├─ 38.085 ALAPScheduleAnalysis.run  ..\..\GitHub\qiskit-ibm-runtime\qiskit_ibm_runtime\transpiler\passes\scheduling\scheduler.py:448
                     │  └─ 37.790 ALAPScheduleAnalysis._visit_block  ..\..\GitHub\qiskit-ibm-runtime\qiskit_ibm_runtime\transpiler\passes\scheduling\scheduler.py:93
                     │     ├─ 35.409 block_order_op_nodes  ..\..\GitHub\qiskit-ibm-runtime\qiskit_ibm_runtime\transpiler\passes\scheduling\utils.py:30
                     │     │  ├─ 31.563 <genexpr>  ..\..\GitHub\qiskit-ibm-runtime\qiskit_ibm_runtime\transpiler\passes\scheduling\utils.py:102
                     │     │  │  ├─ 30.716 _emit  ..\..\GitHub\qiskit-ibm-runtime\qiskit_ibm_runtime\transpiler\passes\scheduling\utils.py:47
                     │     │  │  │  ├─ 18.406 DAGCircuit.is_predecessor  ..\..\GitHub\qiskit-terra\qiskit\dagcircuit\dagcircuit.py:1671
                     │     │  │  │  │  ├─ 12.454 [self]  ..\..\GitHub\qiskit-terra\qiskit\dagcircuit\dagcircuit.py
                     │     │  │  │  │  └─ 5.952 PyDAG.has_edge  <built-in>
                     │     │  │  │  ├─ 10.339 [self]  ..\..\GitHub\qiskit-ibm-runtime\qiskit_ibm_runtime\transpiler\passes\scheduling\utils.py
                     │     │  │  │  └─ 1.325 _is_grouped_measure  ..\..\GitHub\qiskit-ibm-runtime\qiskit_ibm_runtime\transpiler\passes\scheduling\utils.py:37
...
```

with most of the time being spent in the calls to [_emit](https://github.com/Qiskit/qiskit-ibm-runtime/blob/main/qiskit_ibm_runtime/transpiler/passes/scheduling/utils.py#L98) when sorting the circuit graph nodes. 

For comparison with the `qiskit` scheduler:

image

### PR DESCRIPTION
**Pull request by dieris**
### Summary

This is a first attempt at improving the performance of the scheduler, may close #1457 

Given the results of profiling, caching some of the functions seem to be an easy first step. 

### Details and comments

Comparing execution time of the circuit in #  with and without caching of the [_emit](https://github.com/Qiskit/qiskit-ibm-runtime/blob/main/qiskit_ibm_runtime/transpiler/passes/scheduling/utils.py#L98) function:

image

Looking at the cache info for e.g. `N=16` we can see that there is indeed a significant number of repeated calls:
```
emit: CacheInfo(hits=552607, misses=1788, maxsize=None, currsize=1788)
```
There's probably a lot more (or differently) that can be done, open to suggestions.

### CODE REVIEW

**Review by samanthavbarron — COMMENTED** [review]
I think this looks good overall, I have one comment, but if addressed/answered I'll approve!

File: qiskit_ibm_runtime/transpiler/passes/scheduling/utils.py [hunk 1]
```diff
@@ -40,10 +41,11 @@ def _is_block_trigger(node: DAGOpNode) -> bool:
         """Does this node trigger the end of a block?"""
         return isinstance(node.op, ControlFlowOp)
 
+    @lru_cache(maxsize=None)
```
[the remark this answers is not in the document]
Comment (dieris) [reply qiskit_ibm_runtime/transpiler/passes/scheduling/utils.py:44]: I agree that setting a max seems safer. The largest I've seen for N=16 above had ~6k entries. I can try and see how the graph above changes (also for larger N) if we set `maxsize` around that value

[the remark this answers is not in the document]
Comment (dieris) [reply qiskit_ibm_runtime/transpiler/passes/scheduling/utils.py]: This is a bit arbitrary and results fluctuate, but limiting the cache does not seem to affect performance at this scale (if anything, it's slower if left to grow unbounded):
image

**Review by samanthavbarron — APPROVED** [review]
LGTM!

**Pull request merged**